docs(lua): add `vim.str_utf_{start,end,pos}` (#24424)
neovim/neovim: runtime/doc/lua.txt

      • {index}      (number)
      • {use_utf16}  any|nil

vim.str_utf_end({str}, {index})                            *vim.str_utf_end()*
    Gets the distance (in bytes) from the last byte of the codepoint (character) that {index} points to.

    Examples: >lua

     -- The character 'æ' are equal to the bytes '\xc3\xa6' (using UTF-8)

     -- Returns 0 because the index is pointing at the last byte of a character
     vim.str_utf_end('æ', 2)

     -- Returns 1 because the index is pointing at the second last byte of a character
     vim.str_utf_end('æ', 1)
<

    Parameters: ~
      • {str}    (string)
      • {index}  (number)

    Return: ~
        (number)

vim.str_utf_pos({str})                                     *vim.str_utf_pos()*
    Gets a list of the starting byte positions of each UTF-8 codepoint in the given string.

    Embedded NUL bytes are treated as terminating the string.

    Parameters: ~
      • {str}  (string)

    Return: ~
        (table)

vim.str_utf_start({str}, {index})                        *vim.str_utf_start()*

>     Gets the byte offset of {index} compared to the starting byte of the
>     codepoint (character) it points to.
> 

Reviewer: ```suggestion
    Gets the distance (in bytes) from the starting byte of the codepoint (character) that {index} points to.

```

      • {index}      (number)
      • {use_utf16}  any|nil

vim.str_utf_end({str}, {index})                            *vim.str_utf_end()*

>     Returns 0 if the index is pointing at the last byte of a character, 1 if
>     index is pointing at the second last byte of a character, and so on.

Reviewer: ```suggestion
    Gets the distance in bytes of {index} from the last byte of the codepoint (character) it points to.
```

      • {index}      (number)
      • {use_utf16}  any|nil

vim.str_utf_end({str}, {index})                            *vim.str_utf_end()*
    Returns 0 if the index is pointing at the last byte of a character, 1 if

>     index is pointing at the second last byte of a character, and so on.

Reviewer: The returned value behaves like a distance or offset. That seems helpful to mention.

    Examples: >lua

     -- The character 'æ' are equal to the bytes '\xc3\xa6' (using UTF-8)

     -- Returns 0 because the index is pointing at the last byte of a character
     vim.str_utf_end('æ', 2)

     -- Returns 1 because the index is pointing at the second last byte of a character
     vim.str_utf_end('æ', 1)
<

    Parameters: ~
      • {str}    (string)
      • {index}  (number)

    Return: ~
        (number)

vim.str_utf_pos({str})                                     *vim.str_utf_pos()*

>     Gets a list of the starting byte position of each char in the given UTF-8
>     encoded string based on the UTF-8 encoding.

Reviewer: why is "based on the UTF-8 encoding" necessary to mention? It's already implied by the previous mention of "UTF-8 encoded string" .

Is this a more correct and less confusing wording:

```suggestion
    Gets a list of the starting byte positions of each UTF-8 codepoint in the given string.
```

    Embedded NUL bytes are treated as terminating the string.

    Parameters: ~
      • {str}  (string)

    Return: ~
        (table)

vim.str_utf_start({str}, {index})                        *vim.str_utf_start()*

>     Returns 0 if the index is pointing at the first byte of a character, -1 if
>     index is pointing at the second byte of a character, and so on.
> 
>     The reason the function returns negative is so the return value could be
>     added the index and get the starting byte index of a character.

Reviewer: This is too coy. Say the actual meaning instead of describing an infinite series.

```suggestion
    Gets the byte offset of {index} compared to the starting byte of the
    codepoint (character) it points to.

    The result can be added to {index} to get the starting byte of
    a character.
```
